teos/src/api/tor.rs

```diff
@@ -10,126 +10,154 @@ use torut::control::UnauthenticatedConn;
+    ) -> Result<(), Error> {
+        let stream = self
+            .connect_tor_cp()
+            .await
+            .map_err(|e| Error::new(ErrorKind::ConnectionRefused, e))?;
+
+        let mut unauth_conn = UnauthenticatedConn::new(stream);
+
+        let pre_auth = unauth_conn
+            .load_protocol_info()
+            .await
+            .map_err(|e| Error::new(ErrorKind::ConnectionRefused, e))?;
+
+        let auth_data = pre_auth
+            .make_auth_data()?
+            .expect("failed to make auth data");
+
+        unauth_conn.authenticate(&auth_data).await.map_err(|_| {
             Error::new(
-                ErrorKind::Other,
-                format!("failed to create onion hidden service: {}", e),
+                ErrorKind::PermissionDenied,
+                "failed to authenticate with Tor",
             )
         })?;
 
-    print_onion_service(key.clone(), onion_port);
-    service_ready.trigger();
+        let mut auth_conn = unauth_conn.into_authenticated().await;
 
-    // NOTE: Needed to keep connection with control port & hidden service running, as soon as we leave
-    // this function the control port stream is dropped and the hidden service is killed
-    loop {
-        sleep(Duration::from_secs(1)).await;
-        if shutdown_signal_tor.is_triggered() {
-            break;
-        }
-    }
+        auth_conn.set_async_event_handler(Some(|_| async move { Ok(()) }));
 
-    auth_conn
-        .del_onion(
-            &key.public()
-                .get_onion_address()
-                .get_address_without_dot_onion(),
-        )
-        .await
-        .unwrap();
-    Ok(())
-}
-
-async fn connect_tor_cp(addr: SocketAddr) -> Result<TcpStream, Error> {
-    let sock = TcpStream::connect(addr).await.map_err(|_| {
-        Error::new(
-            ErrorKind::ConnectionRefused,
-            "failed to connect to Tor control port",
-        )
-    })?;
-    Ok(sock)
-}
+        auth_conn
+            .add_onion_v3(
+                &self.sk,
+                false,
+                false,
+                false,
+                None,
+                &mut [(self.onion_port, self.api_endpoint)].iter(),
+            )
+            .await
+            .map_err(|e| {
+                Error::new(
+                    ErrorKind::Other,
+                    format!("failed to create onion hidden service: {}", e),
+                )
+            })?;
+
+        log::info!(
+            "Onion service: {}:{}",
+            self.get_onion_address(),
+            self.onion_port
+        );
+        service_ready.trigger();
+
+        // NOTE: Needed to keep connection with control port & hidden service running, as soon as we leave
+        // this function the control port stream is dropped and the hidden service is killed
+        loop {
+            sleep(Duration::from_secs(1)).await;
+            if shutdown_signal_tor.is_triggered() {
+                break;
+            }
+        }
```

Comment:
This can be replace with
```rust
shutdown_signal_tor.await;
```

Reply:
Indeed, nice one! I never liked that chuck of code